de.metas.business/src/main/java/org/adempiere/impexp/inventory/MInventoryImportTableSqlUpdater.java

```diff
@@ -0,0 +1,238 @@
+				.append("FROM (SELECT I_Inventory_ID, d.warehouseValue, d.locatorValue, d.locatorX, d.locatorY, d.locatorZ, d.locatorX1 ")
+				.append("	FROM I_Inventory as inv")
+				.append("	JOIN extractLocatorDimensions(inv.locatorvalue) as d on d.locatorvalue=inv.locatorvalue")
+				.append(") AS dimensions ")
+				.append("WHERE I_IsImported<>'Y' ")
+				.append(whereClause);
+		int no = DB.executeUpdate(sql.toString(), ITrx.TRXNAME_ThreadInherited);
+		logger.debug("Set Warehouse from Locator =" + no);
+	}
+
+	private void dbUpdateWarehouse(@NonNull final String whereClause)
+	{
+		// Set M_Warehouse_ID
+		final StringBuilder sql = new StringBuilder("UPDATE I_Inventory i ")
+				.append("SET M_Warehouse_ID=(SELECT M_Warehouse_ID FROM M_Warehouse w WHERE i.warehouseValue=w.value) ")
+				.append("WHERE M_Warehouse_ID IS NULL ")
+				.append("AND I_IsImported<>'Y' ")
+				.append(whereClause);
+		int no = DB.executeUpdate(sql.toString(), ITrx.TRXNAME_ThreadInherited);
+		logger.debug("Set Warehouse =" + no);
+	}
+
+	private void dbUpdateCreateLocators(@NonNull final String whereClause)
+	{
+		dbUpdateLocators(whereClause);
+		dbCreateLocators();
+	}
+
+	private void dbUpdateLocators(@NonNull final String whereClause)
+	{
+		StringBuilder sql = new StringBuilder("UPDATE I_Inventory i ")
+				.append("SET M_Locator_ID=(SELECT MAX(M_Locator_ID) FROM M_Locator l ")
+				.append("WHERE i.LocatorValue=l.Value AND i.AD_Client_ID=l.AD_Client_ID) ")
+				.append("WHERE M_Locator_ID IS NULL AND LocatorValue IS NOT NULL ")
+				.append("AND I_IsImported<>'Y' ")
+				.append(whereClause);
+		int no = DB.executeUpdate(sql.toString(), ITrx.TRXNAME_ThreadInherited);
+		logger.debug("Set Locator from Value =" + no);
+	}
+
+	private void dbCreateLocators()
+	{
+		final List<I_I_Inventory> unmatchedLocator = Services.get(IQueryBL.class).createQueryBuilder(I_I_Inventory.class)
+				.addOnlyActiveRecordsFilter()
+				.addEqualsFilter(I_I_Inventory.COLUMNNAME_I_IsImported, false)
+				.addNotNull(I_I_Inventory.COLUMN_LocatorValue)
+				.addNotNull(I_I_Inventory.COLUMN_WarehouseValue)
+				.addEqualsFilter(I_I_Inventory.COLUMNNAME_M_Locator_ID, null)
+				.create()
+				.list(I_I_Inventory.class);
+
+		unmatchedLocator.forEach(importRecord -> {
+			final I_M_Locator locator = createNewMLocator(importRecord);
+			importRecord.setM_Locator(locator);
+			InterfaceWrapperHelper.save(importRecord);
+		});
+	}
+
+	private I_M_Locator createNewMLocator(@NonNull final I_I_Inventory importRecord)
+	{
+		final I_M_Locator locator = InterfaceWrapperHelper.newInstance(I_M_Locator.class);
+		locator.setAD_Org_ID(importRecord.getAD_Org_ID());
+		locator.setM_Warehouse_ID(importRecord.getM_Warehouse_ID());
+		locator.setValue(importRecord.getLocatorValue());
+		locator.setX(importRecord.getX());
+		locator.setY(importRecord.getY());
+		locator.setZ(importRecord.getZ());
+		locator.setX1(importRecord.getX1());
+		locator.setDateLastInventory(importRecord.getDateLastInventory());
```

Comment:
what's DateLastInventory and why setting it only when the locator is created?

Reply:
hm.. good point; I need to update also